### PR TITLE
feat(kuma-dp) override bootstrap version

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Summary

This PR introduces an option to override Bootstrap Version on Kuma DP side.

The regular case is that Kuma CP decides which Bootstrap Version (and API Version) it will serve to the DP.

Envoy 1.17 technically drops V2 support, but you can still use it when you specify explicitly `--bootstrap-verison 2` as CLI args to Envoy.
Kuma DP did not know about Bootstrap version so far, so I added extra header that we return from the Kuma CP so Kuma DP can pass the version to Envoy.

Additionally, there are some cases when you would like to override Bootstrap Version for one Kuma DP (for example to V3, while keeping the default Bootstrap to V2 for all other DPs). Those use cases are
1) Testing. When testing, it's usefull to just switch back and forth between versions without restarting CP.
2) Migrations. When running on prod, you may want to check only one DP on V3 that everything is fine while keeping every other on V2.

### Documentation

- This can be included in upgrade guide as a tip when we will change V2 -> V3 by default.

### CODE REVIEW

**Review by nickolaev — APPROVED** [review]
Some nits, otherwise LGTM

File: pkg/config/app/kuma-dp/testdata/default-config.golden.yaml [hunk 1]
```diff
@@ -8,5 +8,6 @@ controlPlane:
 dataplane:
   mesh: default
   drainTime: 30s
+  bootstrapVersion: ""
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/config/app/kuma-dp/testdata/default-config.golden.yaml:11]: the default should be empty. By default, it's control plane which is responsible for picking a version

File: pkg/xds/bootstrap/handler.go [hunk 2]
```diff
@@ -64,6 +64,7 @@ func (b *BootstrapHandler) Handle(resp http.ResponseWriter, req *http.Request) {
 
 	resp.WriteHeader(http.StatusOK)
 	resp.Header().Set("content-type", "text/x-yaml")
+	resp.Header().Set(types.BootstrapVersionHeader, string(version))
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/xds/bootstrap/handler.go:67]: gRPC is overkill in my opionion. Simple REST service with JSON with fields "version" and "bootstrap config" is probably a better idea of those two, but this brings even more problems with backward compatibility, so unless we need more control I'd stick to this for now.

File: pkg/config/app/kuma-dp/testdata/default-config.golden.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (nickolaev) [reply pkg/config/app/kuma-dp/testdata/default-config.golden.yaml:11]: Got it, makes sense.

File: pkg/xds/bootstrap/generator.go [hunk 3]
```diff
@@ -226,14 +228,33 @@ func (b *bootstrapGenerator) verifyAdminPort(adminPort uint32, dataplane *core_m
...
+		cfg, err := b.configForParametersV3(params)
+		if err != nil {
+			return nil, "", err
+		}
+		return cfg, types.BootstrapV3, nil
 	default:
-		return nil, errors.Errorf("invalid API Version %s", b.config.APIVersion)
+		return nil, "", InvalidBootstrapVersion
 	}
 }
```
Comment (bartsmykla) [inline pkg/xds/bootstrap/generator.go:259]: Maybe:
```go
func (b *bootstrapGenerator) configForParameters(params configParameters, version types.BootstrapVersion) (proto.Message, types.BootstrapVersion, error) {
	switch {
	// V2
	case version == types.BootstrapV2:
		fallthrough
	case version == "" && b.config.APIVersion == envoy_common.APIV2: // if client did not overridden bootstrap version, provide bootstrap based on Kuma CP config
		cfg, err := b.configForParametersV2(params)
		if err != nil {
			return nil, "", err
		}
		return cfg, types.BootstrapV2, nil
	// V3
	case version == types.BootstrapV3:
		fallthrough
	case version == "" && b.config.APIVersion == types.BootstrapV3: // if client did not overridden bootstrap version, provide bootstrap based on Kuma CP config
		cfg, err := b.configForParametersV3(params)
		if err != nil {
			return nil, "", err
		}
		return cfg, types.BootstrapV3, nil
	default:
		return nil, "", InvalidBootstrapVersion
	}
}
```
?

File: pkg/xds/bootstrap/handler.go [hunk 4]
```diff
@@ -80,6 +81,14 @@ func handleError(resp http.ResponseWriter, err error, logger logr.Logger) {
 		}
 		return
 	}
+	if err == InvalidBootstrapVersion {
+		resp.WriteHeader(http.StatusBadRequest)
+		_, err = resp.Write([]byte(err.Error()))
+		if err != nil {
+			logger.Error(err, "Error while writing the response")
+		}
```
Comment (bartsmykla) [inline pkg/xds/bootstrap/handler.go:89]:
```suggestion
		if _, err := resp.Write([]byte(err.Error())); err != nil {
			logger.Error(err, "Error while writing the response")
		}
```

**Review by bartsmykla — APPROVED** [review]

**Pull request merged**